src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs

```
@@ -0,0 +1,281 @@
+            // FresherThan(y, x)
+            if (heartbeatReplyNotReceived ||
+                TopologyVersion.IsStalerThanOrEqualTo(currentDescription.TopologyVersion, e.NewServerDescription.TopologyVersion))
+            {
+                SetDescription(e.NewServerDescription);
+            }
+        }
+
+        private void SetDescription(ServerDescription newDescription)
+        {
+            var oldDescription = Interlocked.CompareExchange(ref _currentDescription, value: newDescription, comparand: _currentDescription);
+            OnDescriptionChanged(sender: this, new ServerDescriptionChangedEventArgs(oldDescription, newDescription));
+        }
+
+        private void OnDescriptionChanged(object sender, ServerDescriptionChangedEventArgs e)
+        {
+            if (e.NewServerDescription.HeartbeatException != null)
+            {
+                ConnectionPool.Clear();
+            }
+
+            // propagate event to upper levels
+            TriggerServerDescriptionChanged(this, e);
+        }
+
+        private bool ShouldInvalidateServer(
+            IConnection connection,
+            Exception exception,
+            ServerDescription description,
+            out TopologyVersion invalidatingResponseTopologyVersion)
+        {
+            if (exception is MongoConnectionException mongoConnectionException &&
+                mongoConnectionException.ContainsTimeoutException)
+            {
+                invalidatingResponseTopologyVersion = null;
+                return false;
+            }
+
+            if (__invalidatingExceptions.Contains(exception.GetType()))
+            {
+                invalidatingResponseTopologyVersion = null;
+                return true;
+            }
+
+            var exceptionsToCheck = new[]
+            {
+                exception as MongoCommandException,
+                (exception as MongoWriteConcernException)?.MappedWriteConcernResultException
```

this line corresponds to these lines in the master: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Servers/Server.cs#L526-L542. In 2 words, we just make the same logic as in ExceptionMapper only because the actual exception can be hidden inside `MongoWriteConcernException.writeConcernError`, I extract it into `MappedWriteConcernResultException` and as a result no reason to analyze exception codes on this level anymore.

However, to save time I'm ok to move this logic into a different ticket